src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs

```
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
         {
-            // clear the old values
-            // and write the new ones
-            var sourcesToWrite = sources.Where(s => !s.IsMachineWide && s.IsPersistable);
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+
+            var isDirty = false;
+            AddDisabledSource(source.Name, shouldSkipSave: false, isDirty: ref isDirty);
+        }
+
+        public void DisablePackageSource(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            var isDirty = false;
+            AddDisabledSource(name, shouldSkipSave: false, isDirty: ref isDirty);
+        }
 
-            var existingSettings = (Settings.GetSettingValues(ConfigurationConstants.PackageSources, isPath: true) ??
-                                    Enumerable.Empty<SettingValue>()).Where(setting => !setting.IsMachineWide).ToList();
-            var minPriority = 0;
+        private void AddDisabledSource(string name, bool shouldSkipSave, ref bool isDirty)
+        {
+            var settingsLookup = GetExistingSettingsLookup();
+            var addedInSameFileAsCurrentSource = false;
 
-            // get lowest priority in existingSetting
-            if (existingSettings.Count > 0)
+            if (settingsLookup.TryGetValue(name, out var sourceSetting))
             {
-                minPriority = existingSettings.Min(setting => setting.Priority);
+                try
+                {
+                    (Settings as Settings).AddOrUpdate(sourceSetting.Origin, ConfigurationConstants.DisabledPackageSources, new AddItem(name, "true"));
+                    isDirty = true;
+                    addedInSameFileAsCurrentSource = true;
+                }
+                // We ignore any errors since this means the current source file could not be edited
+                catch { }
```

Similar to the above, why swallow the exception here?